### ISSUE TITLE
False positive for react/jsx-no-leaked-render

### ISSUE THREAD

**Issue by Robloche** [jsx-eslint/eslint-plugin-react#3292]
Following code triggers `react/jsx-no-leaked-render` although it's perfectly valid:

```
return <MyReactComponent
          isDisabled={foo && bar === 0}
          onClick={handleOnClick} />;
```

with:
```
foo: boolean;
bar: number;
```

**Comment by ljharb** [comment]
cc @Belco90

**Comment by Belco90** [comment]
@ljharb thanks for pinging! I can take care of this, feel free to assign it to me.

**Comment by Belco90** [comment]
@Robloche regarding the false positive reported: I think sometimes you would be interested into having this rule applied to props also, like:

```jsx
<MyReactComponent
  header={someNumber && <Header />}
  isDisabled={foo && bar === 0}
/>
```

So I think the proper fix for this would be to make sure the right side of the && operator is a condition or not.

**Comment by Robloche** [comment]
Agreed. So what's you recommendation?

As for now, I simply rewrote my code, adding a local `isDisabled` boolean.
And since this makes the code clearer, I guess it's a good solution.

**Comment by Belco90** [comment]
> The right side should be able to be any value that’s safe to render, which includes booleans.

That's true, but now I don't know what to do with this false positive since I can't figure the types of the operators. Any suggestions?

**Comment by ljharb** [comment]
The RHS is an === comparison; it can’t ever be anything but a boolean. You don’t need to know any types to figure that part out.

**Comment by Belco90** [comment]
> The RHS is an === comparison; it can’t ever be anything but a boolean. You don’t need to know any types to figure that part out.

Exactly, so since the RHS is something renderable, this rule should complain about it since we don't know the type for the LHS (it could be a zero, so the rule reported it as expected).

I was suggesting to report props only when there is a && operator and the RHS is a component so we can assume you want to pass that prop to render something, rather than just use the value for logic purposes.

**Comment by Robloche** [comment]
Yes, I'm aligned with your suggestion. That should do the trick.

**Comment by Belco90** [comment]
This could be even fixed by introducing a new option: `reportProps`. It could receive the following values:

- `"always"`: as it is now
- `"components"`: when the RHS is a component, as suggested
- `"never"`: so no props are reported

**Comment by ljharb** [comment]
In the case of https://github.com/jsx-eslint/eslint-plugin-react/issues/3292#issuecomment-1132947856, with no type information, obviously both of those should warn.

**Comment by texonidas** [comment]
I'm also getting a very incorrect error for this:

```js
const DISPLAY = {
   WELCOME: 'WELCOME',
   FOO: 'FOO'
};

const Component = () => {
    const [display, setDisplay] = useState(DISPLAY.WELCOME);
    
    return (
        <div>
            {(!display || display === DISPLAY.WELCOME) && (
                 <span>my condition produces a linting error even though I can't not be boolean</span>
            )}
        </div>
    );
};
```

**Comment by ljharb** [comment]
In that case, since `!display` and `display === something` are booleans, there shouldn't be any TS required to statically know the condition is safe (the runtime typeof is irrelevant)

**Comment by Belco90** [comment]
@ljharb I'll need way more time to try checking the TS types when available for the operators of the logical expression as discussed in this issue, since I've never done this before.

**Comment by fabiendem** [comment]
Not sure it can help but I have been using this relatively simple rule so far https://github.com/grailed-code/eslint-plugin-grailed/blob/main/lib/rules/jsx-no-ampersands.js
Tests: https://github.com/grailed-code/eslint-plugin-grailed/blob/main/tests/lib/rules/jsx-no-ampersands.js
It doesn't raise the false positives, even without any knowledge of the types.

**Comment by pke** [comment]
A new option would be good. Props should not be validated by default.

Also the fix should not result to `null` in ternary operator but to `undefined` or it should be configurable. `null` according to Crockford should not be used: if you want a falsy value use `undefined`, because `null` does not work with default params and therefore completely changes the code flow.

**Comment by Belco90** [comment]
> Also the fix should not result to `null` in ternary operator but to `undefined` or it should be configurable. `null` according to Crockford should not be used: if you want a falsy value use `undefined`, because `null` does not work with default params and therefore completely changes the code flow.

Returning `null` is required since prior to React v18 returning `undefined` is considered as a missed return statement. React itself will complain with a message like: _Uncaught Error: Error(...): Nothing was returned from render. This usually means a return statement is missing. Or, to render nothing, return null._

**Comment by pke** [comment]
re: `null` thats correct for render returns but not for props. I was unclear. I meant for props fixes it should use undefined and not null.

**Comment by Belco90** [comment]
@pke I see, I guess we can discuss on how to improve the rule reporting on props other than `children`. Not really sure what to do with it to be honest, apart from the option to decide between `null` or `undefined`.

**Comment by pke** [comment]
Given those 2 components:

```tsx
function Button({ type = "primary"}) {
 ...
}

function Parent({ count: number }) {
  return <Button type={count && "secondary"}/>
}
```

the linting rule would change this:

```diff
-return <Button type={count && "secondary"}/>
+return <Button type={count ? "secondary": null }/>
```

And when it sends `null` for `type` to the `<Button>` the Buttons default assignment for the `type` prop does not work anymore.

**Comment by ljharb** [comment]
"`null` should not be used" is useless and antiquated advice that the larger JS ecosystem has resoundingly rejected, so let's just throw that out the window right now.

**Comment by pke** [comment]
You see the problem with null in that case though?

**Comment by ljharb** [comment]
Yes, but a present undefined prop isn't the same thing as an absent prop either.

In this specific case, if the prop value is `foo && bar === 0` then the rule simply shouldn't be checking it - so yes, I do agree that this rule should not be checking prop values that aren't "`children` on an HTML element".

**Comment by ljharb** [comment]
Like for the props object the component receives. `<Foo />` and `<Foo x={undefined} />` are different - the former gets a props object of `{}`, the latter `{ x: undefined }`, and these are observably different.

**Comment by Belco90** [comment]
> Has typescript awareness landed for this rule?

@shamilovtim I'm afraid it didn't, and I'm not really available to handle this for now.

**Comment by Jackman3005** [comment]
> > Has typescript awareness landed for this rule?
> 
> @shamilovtim I'm afraid it didn't, and I'm not really available to handle this for now.

That's a bummer :/ I imagine it would be quite a bit easier to determine if the variable is a boolean in a Typescript aware context.

**Comment by Belco90** [comment]
@levrik IMO we should include an option in the rule to avoid reporting props other than `children` (enabled by default to avoid changing the default behavior I guess).

**Comment by ljharb** [comment]
@levrik i believe that will be fixed in the next release.

**Comment by ljharb** [comment]
@levrik have you tried the latest version?

**Comment by levrik** [comment]
@ljharb I tested with version 7.31.8

**Comment by aleclarson** [comment]
I've fixed this in #3441 (new `ignoreAttributes` option)

Can one of you write the docs for that PR? Just type them in this issue, and I'll copy them into my PR. Thanks!

Of course, the best solution would involve TypeScript type-checking. Does this plugin allow that, or maybe someone can provide guidance here?

**Comment by aleclarson** [comment]
Oh I just found this plugin, which uses TypeScript:
https://www.npmjs.com/package/eslint-plugin-jsx-expressions

**Comment by karlhorky** [comment]
> Oh I just found this plugin, which uses TypeScript:
> [npmjs.com/package/eslint-plugin-jsx-expressions](https://www.npmjs.com/package/eslint-plugin-jsx-expressions)

Oh nice, looks like a nice plugin!

@hpersson since you're the author of this, would you be interested in contributing TypeScript type checking also to `react/jsx-no-leaked-render` rule in `eslint-plugin-react`?

**Comment by tcl333** [comment]
Why was the `ignoreAttributes` option reverted? This rule is unusable for me now.

**Comment by tcl333** [comment]
This rule is unusable to me until PR #3441 is landed. I need to handle a very basic usecase of `<Component isFooBar={isFoo && isBar}>`. This rule currently auto-breaks that line by converting it into `<Component isFooBar={isFoo ? isBar : null }>`. My component doesn't take `null` as a prop, nor should it be required to.

**Comment by shamilovtim** [comment]
Yeah we can't use this rule either. It's not inferring bools. One of these days I'll be on a project where I can afford to contribute back here. Thank you to the maintainers and contributors.

**Comment by danielrotaermel** [comment]
> Oh I just found this plugin, which uses TypeScript:
> https://www.npmjs.com/package/eslint-plugin-jsx-expressions

Unfortunately the plugin does not work with `@typescript-eslint/parser` `^6` right now.
But there is a an open issue about it https://github.com/hluisson/eslint-plugin-jsx-expressions/issues/13

**Comment by karlhorky** [comment]
We're on the latest version by overriding the package version installed, eg using [npm Overrides](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#overrides):

`package.json`

```json
{
  "overrides": {
    "@typescript-eslint/eslint-plugin": "6.9.1",
    "@typescript-eslint/parser": "6.9.1"
  }
}
```

(or [Yarn Resolutions](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/) or [pnpm Overrides](https://pnpm.io/package_json#pnpmoverrides))

**Comment by ljharb** [comment]
Using overrides will cause a number of rules to be broken; peer dep warnings need to be respected.

**Comment by karlhorky** [comment]
Yeah, good warning for many cases. I was talking about a workaround for this specific case, where it does not cause breakage in our experience.

**Comment by ljharb** [comment]
If that were true, we wouldn’t need #3629 ¯\\\_(ツ)\_/¯

**Comment by iway1** [comment]
makes no sense that this checked props in the first place

**Comment by kcsujeet** [comment]
It's 2025 and I still have to do `Boolean(showButton)` even though `showButton` is a boolean. *sigh

**Comment by kachkaev** [comment]
I recently switched to [`@eslint-react/no-leaked-conditional-rendering`](https://eslint-react.xyz/docs/rules/no-leaked-conditional-rendering). It is TypeScript-aware and works like a charm ✨

**Comment by tms-mayur-padshala** [comment]
> Agreed. So what's you recommendation?
> 
> As for now, I simply rewrote my code, adding a local `isDisabled` boolean. And since this makes the code clearer, I guess it's a good solution.

You can aslo do 
```js
return <MyReactComponent
          isDisabled={Boolean(foo && bar === 0)}
          onClick={handleOnClick} />;
```
